### PR TITLE
Add fetching users from keycloak rest-api

### PR DESCRIPTION
**Pull request by ewelinagr**

### CODE REVIEW

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        MultiValueMap<String, String> body = new LinkedMultiValueMap<>()
+        body.add("grant_type", "refresh_token")
+        body.add("client_id", clientId)
+        body.add("scope", "offline_access")
+        body.add("refresh_token", offlineToken)
+
+        def url = URI.create("$keycloakServerUrl/realms/$realm/protocol/openid-connect/token")
+        HttpEntity<?> httpEntity = new HttpEntity<Object>(body, headers)
+        ResponseEntity<Object> response = new RestTemplate().exchange(url, HttpMethod.POST, httpEntity, Object.class)
```
Comment (ewelinagr) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy]: I do not like this part (using RestTemplate here), but did not find other way to get the offline access token for CustomClientRequestFactory

File: transmart-api-server/grails-app/conf/spring/resources.groovy [hunk 2]
```diff
@@ -1,6 +1,8 @@
+import org.transmartproject.api.server.client.CustomClientRequestFactory
 import org.transmartproject.api.server.user.KeycloakUserResourceService
 
 beans = {
     requestContextListener(org.springframework.web.context.request.RequestContextListener)
     userResource(KeycloakUserResourceService)
+    customClientRequestFactory(CustomClientRequestFactory)
```
Comment (forus) [inline transmart-api-server/grails-app/conf/spring/resources.groovy]: We don't need that. Do we?

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.http.client.ClientHttpRequestFactory
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
+import org.springframework.stereotype.Component
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.util.MultiValueMap
+import org.springframework.web.client.RestTemplate
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+class CustomClientRequestFactory extends HttpComponentsClientHttpRequestFactory implements ClientHttpRequestFactory {
```
Comment (forus) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy]: Ah. I was confused with naming above. Could it be renamed to `OfflineTokenRequestFactory` or smth.?

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        String offlineToken = getOfflineAccessToken()
+        request.setHeader(AUTHORIZATION_HEADER, "Bearer " + offlineToken)
+    }
+
+    /**
+     * Get access token from Keycloak based on the client_id and the offline token, which is stored in the config.
+     * Offline token is a type of a classic Refresh token, but it never expires.
+     * @return access token
+     */
+    private String getOfflineAccessToken(){
```
Comment (forus) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy]: AFAIK offline token gives us right to get a new access token. `getNewAccessTokenByOfflineTokenAndClientId()` express the purpose of the method better, more mouthful though.

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Value('${keycloak.serverUrl}')
+    private String keycloakServerUrl
+
+    public static final String AUTHORIZATION_HEADER = "Authorization"
+
+
+    @Override
+    protected void postProcessHttpRequest(HttpUriRequest request) {
+        String offlineToken = getOfflineAccessToken()
```
Comment (forus) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy]: it's not the offline token. We might apply some caching here at some point: fetch new access token only when the current one has expired. For 2 rest api calls it's not that important though.

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomRestTemplate.groovy [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package org.transmartproject.api.server.client
+
+import org.springframework.web.client.RestOperations
+import org.springframework.web.client.RestTemplate
+
+class CustomRestTemplate extends RestTemplate implements RestOperations {
```
Comment (forus) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomRestTemplate.groovy]: We don't have to have a custom class. Couldn't it be matter composing in the spring configuration like below:
```
@Bean
RestTemplate offlineTokenBasedRestTemplate(CustomClientRequestFactory factory) {
 new RestTemplate(factory)
}
```

File: transmart-api-server/src/test/groovy/org/transmartproject/api/server/user/KeycloakUserResourceServiceSpec.groovy [hunk 7]
```diff
@@ -68,18 +76,90 @@ class KeycloakUserResourceServiceSpec extends Specification {
...
+                                "client2": new ClientMappingsRepresentation(
+                                        mappings: [])
+                        ]
+                )]
+
+        ResponseEntity userResponse = new ResponseEntity(keycloakMockUsers, HttpStatus.OK)
+        ResponseEntity user1RolesResponse = new ResponseEntity(keycloakMockUser1Roles, HttpStatus.OK)
+        ResponseEntity user2RolesResponse = new ResponseEntity(keycloakMockUser2Roles, HttpStatus.OK)
+
+        def restTemplate = Mock(CustomRestTemplate, {
```
Comment (forus) [inline transmart-api-server/src/test/groovy/org/transmartproject/api/server/user/KeycloakUserResourceServiceSpec.groovy]: Mock `RestOperations` interface. Then no `cglib` library needed.

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy (hunk 1 above)
Comment (forus) [reply transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy]: You might want to have a field of `RestOperations` type instead and use spring injection. Watch out to provide right rest template instance, not managed with the offline token to escape amazing cycle).

File: transmart-api-server/src/main/groovy/org/transmartproject/api/server/user/KeycloakUserResourceService.groovy [hunk 8]
```diff
@@ -20,19 +25,38 @@ class KeycloakUserResourceService implements UsersResource {
     @Autowired
     LegacyAuthorisationChecks authorisationChecks
 
+    @Autowired
+    CustomRestTemplate restTemplate
```
Comment (forus) [inline transmart-api-server/src/main/groovy/org/transmartproject/api/server/user/KeycloakUserResourceService.groovy]: Use field of `RestOperations` type instead.

**Pull request merged**